src/System.IO/src/System/IO/StreamReader.cs

```diff
@@ -526,7 +526,10 @@ private void DetectEncoding()
                 _byteBuffer[2] == 0xFE && _byteBuffer[3] == 0xFF)
             {
                 // Big Endian UTF32
-                _encoding = new UTF32Encoding(bigEndian: true, byteOrderMark: true);
+                // Unfortunately, there is no public Encoding.BigEndianUTF32 property, so instead
+                // we call Encoding.GetEncoding, which will retrieve a cached instance.
+                const int CodePageUTF32BE = 12001;
+                _encoding = Encoding.GetEncoding(CodePageUTF32BE);
```

Comment:
Yes, that's a good point, I hadn't thought of that...

Should we be concerned about other existing callers of `Encoding.GetEncoding(int/string)` that may get back an unexpected encoding?

For example:
 - `System.Net.Http.Headers.UriHeaderParser` [calls `Encoding.GetEncoding("utf-8", ...)`](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Net.Http/src/System/Net/Http/Headers/UriHeaderParser.cs#L101)
 - There's code throughout `System.Net.Mail` that makes similar [calls to `Encoding.GetEncoding("utf-8");`](https://github.com/dotnet/corefx/blob/abc007cbd6ef6180f1e044547d6441058122203a/src/System.Net.Mail/src/System/Net/Mail/MailAddress.cs#L76)
 - `System.Data.SqlTypes.SqlXml` [calls `Encoding.GetEncoding("utf-16");`](https://github.com/dotnet/corefx/blob/c5c75db7c00047d603522fd6e21f350099e09bbb/src/System.Data.Common/src/System/Data/SQLTypes/SqlXml.cs#L195)
 - etc.

Should `Encoding.GetEncoding` be changed to check for and return the handful of known encodings up-front before deferring to the providers?